src/api/context/context.zig

```diff
@@ -0,0 +1,430 @@
+        const new_map_ptr = try allocator.create(EntryMap);
+        if (self.entries) |existing| {
+            new_map_ptr.* = try existing.clone(allocator);
+        } else {
+            new_map_ptr.* = EntryMap{};
+        }
+        errdefer allocator.destroy(new_map_ptr);
+        try new_map_ptr.put(allocator, key, value);
+        return Self{ .entries = new_map_ptr, .allocator = allocator };
+    }
+
+    pub fn deinit(self: *Self) void {
+        if (self.entries) |entries| {
+            if (self.allocator) |allocator| {
+                entries.deinit(allocator);
+                allocator.destroy(entries);
+            }
+        }
+        self.* = init();
+    }
+};
+
+/// Token representing a position in the context stack.
+///
+/// Tokens are returned by `attachContext()` and must be used with `detachContext()`
+/// to maintain proper stack ordering. Tokens implement the LIFO (last-in, first-out)
+/// semantics required by the OpenTelemetry specification.
+pub const Token = struct {
+    position: u32,
+};
+
+/// Thread-local context stack for implicit context propagation.
+///
+/// The context stack enables automatic context propagation without explicit
+/// parameter passing. This follows the OpenTelemetry pattern where the "current"
+/// context is implicitly available throughout the call stack.
+const ContextStack = struct {
+    /// Stack of attached contexts
+    contexts: std.ArrayList(Context),
+
+    const Self = @This();
+
+    /// Initializes a new context stack.
+    fn init(allocator: std.mem.Allocator) Self {
+        return Self{ .contexts = std.ArrayList(Context).init(allocator) };
+    }
+
+    /// Releases all resources and contexts in the stack.
+    fn deinit(self: *Self) void {
+        for (self.contexts.items) |*ctx| {
+            ctx.deinit();
+        }
+        self.contexts.deinit();
+    }
+
+    /// Returns the current (top-most) context from the stack.
+    fn current(self: *const Self) Context {
+        return if (self.contexts.items.len > 0) self.contexts.items[self.contexts.items.len - 1] else Context.init();
+    }
+
+    /// Pushes a context onto the stack and returns a detachment token.
+    fn attach(self: *Self, context: Context) !Token {
+        const position = @as(u32, @intCast(self.contexts.items.len));
+        try self.contexts.append(context);
+        return Token{ .position = position };
+    }
+
+    /// Removes the top context from the stack if it matches the token position.
+    fn detach(self: *Self, token: Token) !bool {
+        const expected_len = token.position + 1;
+        if (expected_len != self.contexts.items.len) {
+            return error.DetachOrderError;
+        }
+        var popped = self.contexts.pop().?;
+        (&popped).deinit();
+        return true;
+    }
+};
+
+/// Thread-local storage for the context stack.
+/// Each thread maintains its own independent context stack.
+threadlocal var context_stack: ?*ContextStack = null;
```

Comment:
Nailed it 👍🏼 
This is the same conclusion @kmos and I reached when reading the Context documentation for the first time last week.